libraries/chain/worker_evaluators.cpp

```diff
@@ -480,4 +468,75 @@ namespace golos { namespace chain {
+        ASSERT_REQ_HF(STEEMIT_HARDFORK_0_21__1013, "worker_payment_approve_operation");
+
+        auto approver_witness = _db.get_witness(o.approver);
+        GOLOS_CHECK_LOGIC(approver_witness.schedule == witness_object::top19,
+            logic_exception::approver_of_payment_should_be_in_top19_of_witnesses,
+            "Approver of payment should be in Top 19 of witnesses");
+
+        GOLOS_CHECK_LOGIC(o.state != worker_techspec_approve_state::approve,
+            logic_exception::payment_does_not_need_approving_it_can_be_only_disapproved,
+            "Payment does not need approving, it can be only disapproved");
+
+        const auto& wro_post = _db.get_comment(o.worker_result_author, o.worker_result_permlink);
+        const auto& wto = _db.get_worker_result(o.worker_result_author, o.worker_result_permlink);
+
+        const auto& wpo_post = _db.get_comment(wto.worker_proposal_author, wto.worker_proposal_permlink);
+        const auto& wpo = _db.get_worker_proposal(wpo_post.id);
+
+        GOLOS_CHECK_LOGIC(wpo.state == worker_proposal_state::payment,
+            logic_exception::worker_proposal_should_be_in_payment_state,
+            "Worker proposal should be in payment state");
+
+        const auto& wpao_idx = _db.get_index<worker_payment_approve_index, by_result_approver>();
+        auto wpao_itr = wpao_idx.find(std::make_tuple(wro_post.id, o.approver));
+
+        if (o.state == worker_techspec_approve_state::abstain) {
+            if (wpao_itr != wpao_idx.end()) {
+                _db.remove(*wpao_itr);
+            }
+
+            return;
+        }
+
+        if (wpao_itr != wpao_idx.end()) {
+            _db.modify(*wpao_itr, [&](worker_payment_approve_object& wpao) {
+                wpao.state = o.state;
+            });
+        } else {
+            _db.create<worker_payment_approve_object>([&](worker_payment_approve_object& wpao) {
+                wpao.approver = o.approver;
+                wpao.post = wro_post.id;
+                wpao.state = o.state;
+            });
+        }
+
+        auto disapprovers = count_worker_approves(_db, wpao_idx, wro_post.id, worker_techspec_approve_state::disapprove);
+
+        if (disapprovers < STEEMIT_SUPER_MAJOR_VOTED_WITNESSES) {
+            return;
+        }
+
+        const auto& gpo = _db.get_dynamic_global_properties();
+        _db.modify(gpo, [&](dynamic_global_property_object& gpo) {
+            gpo.worker_consumption_per_month -= wto.month_consumption;
+        });
+
+        _db.modify(wpo, [&](worker_proposal_object& wpo) {
+            if (wpo.type == worker_proposal_type::premade_work) {
+                wpo.state = worker_proposal_state::created;
+                return;
+            }
+
+            wpo.state = worker_proposal_state::work;
+        });
+
+        _db.modify(wto, [&](worker_techspec_object& wto) {
+            wto.next_cashout_time = time_point_sec::maximum();
+            wto.month_consumption = asset(0, STEEM_SYMBOL);
```

Comment:
Lets add issue to extend `worker_techspec_state` with extra states.

Reply:
@afalaleev Fixed https://github.com/GolosChain/golos/pull/1137